### DISCUSSION TITLE
How to escape json schema `defaultSnippets` snippet syntax

### DISCUSSION THREAD

**Discussion by RedCMD** [microsoft/vscode-discussions#984] [Extension Development QnA] [answered]
I've been making a [schema](https://github.com/RedCMD/TmLanguage-Syntax-Highlighter/blob/main/vscode.tmLanguage.schema.json) for my extension and wanted to add some `"defaultSnippets"`
which worked fine until I tried to include the literal characters `$1`
it comes out as a tabstop as expected
but when I went to escape it, I got unexpected results and wasn't able to find an escaping solution

minimal schema:
```json
{
	"$schema": "http://json-schema.org/draft-07/schema#",
	"title": "defaultSnippets",
	"type": "object",
	"properties": {
		"vscode bug": {
			"defaultSnippets": [
				{
					"label": "vscode bug",
					"body": [
						"$1",
						"\\$2",
						"\\\\$3",
						"$\\4",
						"$\\\\5",
						"\\$\\6"
					]
				}
			]
		}
	}
}
```

minimal json file with schema active:
![image](https://github.com/microsoft/vscode-discussions/assets/33529441/0503d409-97b7-490e-a35e-a401634cfbc3)

trigger intellisense code completions with ctrl+space:
![image](https://github.com/microsoft/vscode-discussions/assets/33529441/dc70d026-9f75-4afa-b350-a41e5b4c6ee1)

unexpected escaping:
![image](https://github.com/microsoft/vscode-discussions/assets/33529441/74fc1eaa-a5e4-4319-90ed-0dc3c89bee22)

[userdefinedsnippets#_grammar](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_grammar) states I can just escape the `$` with a backslash, but it doesn't work

**Comment by isidorn** [comment]
Maybe @aeschli can help here

**Comment by aeschli** [comment]
The problem is that the item is stringified, so all `\` become `\\`.

The solution is to use `bodyText`. There the string is taken as-is (not strigified)
```
{
	"$schema": "http://json-schema.org/draft-07/schema#",
	"title": "defaultSnippets",
	"type": "object",
	"properties": {
		"vscode bug": {
			"defaultSnippets": [
				{
					"label": "vscode bug",
					"bodyText": "[ \"\\$1\" ]"

				}
			]
		}
	}
}
```

**Reply by RedCMD** [reply]
the problem with `bodyText` is that it doesn't auto format the resulting snippet when inserted

however `body` correctly formats the snippet based on the user's indent settings etc

**Comment by aeschli** [comment, marked as answer]
One more trick (my memory comes back):
In `body`, start your string literal with `^`. Again, that string literal will not be stringified but inserted as is
```
{
	"$schema": "http://json-schema.org/draft-07/schema#",
	"title": "defaultSnippets",
	"type": "object",
	"properties": {
		"vscode bug": {
			"defaultSnippets": [
				{
					"label": "vscode bug",
					"body": [
						"^\"\\$1\""
					]
				}
			]
		}
	}
}
```

**Reply by RedCMD** [reply]
Thank you
that works well for json `values`
however it doesn't work for json `keys`

```json
"body": {
	"^\\$1": "^\"\\$1\"",
	"$schema": "https://raw.githubusercontent.com/RedCMD/TmLanguage-Syntax-Highlighter/main/vscode.tmLanguage.schema.json",
	"^\\$schema": "https://raw.githubusercontent.com/RedCMD/TmLanguage-Syntax-Highlighter/main/vscode.tmLanguage.schema.json",
	"$${_///}schema": "https://raw.githubusercontent.com/RedCMD/TmLanguage-Syntax-Highlighter/main/vscode.tmLanguage.schema.json"
}
```
![image](https://github.com/microsoft/vscode-discussions/assets/33529441/02c5e785-9994-4b85-8d5d-5ca4dd1dd56a)
